src/software/util/make_enum/make_enum_helpers.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <array>
+#include <string_view>
+#include <utility>
+
+/**
+ * Removes all leading and trailing whitespace from a string.
+ *
+ * @param str the string to strip whitespace from
+ *
+ * @return str stripped of all its leading and trailing whitespace
+ */
+constexpr std::string_view stripWhitespace(const std::string_view str)
+{
+    constexpr std::string_view WHITESPACE = " \t\n\f\r\v";
+
+    const size_t str_begin = str.find_first_not_of(WHITESPACE);
+    if (str_begin == std::string::npos)
+    {
+        return "";
+    }
+
+    const size_t str_end   = str.find_last_not_of(WHITESPACE);
+    const size_t str_range = str_end - str_begin + 1;
+
+    return str.substr(str_begin, str_range);
+}
+
+/**
+ * Counts the number of substrings separated by a delimiter.
+ *
+ * @param str a string of substrings separated by the delimiter
+ * @param delimiter the delimiting character between substrings
+ *
+ * @return the number of substrings in the string
+ */
+constexpr size_t countSubstrings(std::string_view str, char delimiter)
+{
+    if (str.empty())
+    {
+        return 0;
+    }
+
+    // There is always at least one substring if the string isn't empty
+    size_t count = 1;
+    for (size_t i = 0; i < str.size(); ++i)
+    {
+        // Trailing commas are not counted
+        if (str.at(i) == ',' && (i != str.size() - 1))
+        {
+            ++count;
+        }
+    }
+    return count;
+}
+
+/**
+ * Splits a string of substrings separated by a delimiter into an array containing
+ * those substrings. All leading and trailing whitespace for each substring will be
+ * removed.
+ *
+ * @param str a string of substrings separated by the delimiter
+ * @param delimiter the delimiting character between substrings
+ * @param index_seq an index sequence 0, 1, 2, ..., (N - 1) where N is the number of
+ * substrings in str (use countSubstrings to count the number of substrings)
+ *
+ * @return an array containing the substrings
+ */
+template <size_t... INDEX_SEQ>
+constexpr auto splitString(std::string_view str, [[maybe_unused]] char delimiter,
+                           std::index_sequence<INDEX_SEQ...> index_seq)
+{
+    std::array<std::string_view, sizeof...(INDEX_SEQ)> result = {};
+    std::string_view remaining                                = str;
+    size_t delimiter_pos                                      = 0;
+
+    // This pattern will be instantiated for each index in index_seq
+    ((delimiter_pos = remaining.find(delimiter),
+
+      // Get the substring before the first delimiter in the remaining string
+      result[INDEX_SEQ] = stripWhitespace(remaining.substr(0, delimiter_pos)),
+
+      // Set remaining string to everything after the first delimiter
+      remaining = (delimiter_pos == std::string_view::npos)
+                      ? ""
+                      : remaining.substr(delimiter_pos + 1)),
+     ...);
```

Comment:
Wow, did not know this was a thing. I got some reading to do.

Does the `((some_code), ...)` (called fold operator?! from what I could find) act as a for loop over the `INDEX_SEQ` sequence? 
And does `result[INDEX_SEQ]` on line 82 act as a way of indexing into `result` for each value in `INDEX_SEQ`? 

So this would be like a metaprogramming(?) way of mapping values from `index_seq` to `result`?

I guess would it be able to achieve the same thing at compile time with a for loop?

Reply:
Yeah, this is a nifty metaprogramming trick I found about recently.

We can't guarantee for-loops will be evaluated/unrolled at compile time, *but* we can iterate over a template parameter pack at compile time (the `size_t... INDEX_SEQ` parameter, which lets the template accept any number of template args of type `size_t`). We iterate over the parameter pack by using the fold operator, which generates code that "expands" out the given pattern for each given arg in the parameter pack, with each instantiation of the pattern separated by the given operator (here I'm using the `,` operator which basically does nothing). 

By using `std::make_index_sequence` to generate 0, 1, 2, ..., N - 1 as template args, this lets us write an unrolled for-loop over range(N) at compile time.